File: src/main/kotlin/FindHelper.java
import model.*;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static java.lang.StrictMath.abs;

public class FindHelper {

    private Wizard wizard;
    private World world;
    private Game game;

    private static Map<List<Boolean>, List<LivingUnit>> allUnitsCache = new HashMap<>();
    private static Map<List<Boolean>, List<LivingUnit>> allMovingUnitsCache = new HashMap<>();
    private static Map<List<Boolean>, List<Wizard>> allWizardsCache = new HashMap<>();
    private static Map<Boolean, List<Building>> allBuldings = new HashMap<>();
    private static Map<List<Boolean>, List<Minion>> allMinions = new HashMap<>();
    private static List<Tree> allTrees = null;

    public static void clearCache() {
        allUnitsCache = new HashMap<>();
        allMovingUnitsCache = new HashMap<>();
        allWizardsCache = new HashMap<>();
        allBuldings = new HashMap<>();
        allMinions = new HashMap<>();
        allTrees = null;
    }

    public FindHelper(World world, Game game, Wizard wizard) {
        this.world = world;
        this.game = game;
        this.wizard = wizard;
    }

    public List<LivingUnit> getAllUnits(boolean withTrees, boolean onlyEnemy, boolean onlyNearest) {
        List<Boolean> cacheKey = Arrays.asList(withTrees, onlyEnemy, onlyNearest);
        if (allUnitsCache.get(cacheKey) != null) return allUnitsCache.get(cacheKey);

        List<LivingUnit> units = new ArrayList<>();

        units.addAll(getAllWizards(onlyEnemy, onlyNearest));
        units.addAll(getAllBuldings(onlyEnemy));
        units.addAll(getAllMinions(onlyEnemy, onlyNearest));

        if (withTrees)
            units.addAll(Arrays.asList(world.getTrees()));

        allUnitsCache.put(cacheKey, units);

        return units;
    }

    public List<LivingUnit> getAllMovingUnits(boolean onlyEnemy, boolean onlyNearest) {
        List<Boolean> cacheKey = Arrays.asList(onlyEnemy, onlyNearest);
        if (allMovingUnitsCache.get(cacheKey) != null) return allMovingUnitsCache.get(cacheKey);

        List<LivingUnit> units = new ArrayList<>();

        units.addAll(getAllWizards(onlyEnemy, onlyNearest));
        units.addAll(getAllMinions(onlyEnemy, onlyNearest));

        allMovingUnitsCache.put(cacheKey, units);

        return units;
    }

    public List<Wizard> getAllWizards(boolean onlyEnemy, boolean onlyNearest) {
        List<Boolean> cacheKey = Arrays.asList(onlyEnemy, onlyNearest);

        if (allWizardsCache.get(cacheKey) != null) return allWizardsCache.get(cacheKey);

        List<Wizard> newUnins = Arrays.stream(world.getWizards())
                .filter(wizard -> !wizard.isMe())
                .filter(filterLivingUnits(onlyEnemy, onlyNearest))
                .collect(Collectors.toList());

        allWizardsCache.put(cacheKey, newUnins);

        return newUnins;
    }

    private Predicate<LivingUnit> filterLivingUnits(boolean onlyEnemy, boolean onlyNearest) {
        return unit -> (!onlyEnemy || isEnemy(wizard.getFaction(), unit))
                && (!onlyNearest || abs(unit.getX() - wizard.getX()) < game.getWizardCastRange() * 3)
                && (!onlyNearest || abs(unit.getY() - wizard.getY()) < game.getWizardCastRange() * 3);

    }

    public List<Building> getAllBuldings(boolean onlyEnemy) {
        Boolean cacheKey = onlyEnemy;

        if (allBuldings.get(cacheKey) != null) return allBuldings.get(cacheKey);

        List<Building> newUnits = Arrays.stream(world.getBuildings())
                .filter(filterLivingUnits(onlyEnemy, false))
                .collect(Collectors.toList());

        allBuldings.put(cacheKey, newUnits);

        return newUnits;
    }

    public List<Minion> getAllMinions(boolean onlyEnemy, boolean onlyNearest) {
        List<Boolean> cacheKey = Arrays.asList(onlyEnemy, onlyNearest);

        if (allMinions.get(cacheKey) != null) return allMinions.get(cacheKey);

        List<Minion> newUnits = Arrays.stream(world.getMinions())
                .filter(filterLivingUnits(onlyEnemy, onlyNearest))
                .collect(Collectors.toList());

        allMinions.put(cacheKey, newUnits);

        return newUnits;
    }

    public List<Tree> getAllTrees() {
        if (allTrees != null) return allTrees;

        List<Tree> newUnits = Arrays.stream(world.getTrees())
                .filter(filterLivingUnits(false, true))
                .collect(Collectors.toList());

        allTrees = newUnits;

        return allTrees;
    }

    public boolean isEnemy(Faction self, LivingUnit unit) {
        return self != unit.getFaction() && unit.getFaction() != Faction.NEUTRAL;
    }

    public Optional<LivingUnit> getNearestEnemy() {
        Optional<LivingUnit> nearestWizard = getNearestTarget(world.getWizards());

        if (nearestWizard.isPresent()) return nearestWizard;

        Optional<LivingUnit> nearestBuilding = getNearestTarget(world.getBuildings());

        if (nearestBuilding.isPresent()) return nearestBuilding;

        else return getNearestTarget(world.getMinions());
    }

    public Optional<LivingUnit> getNearestTarget(LivingUnit[] targets) {
        List<LivingUnit> nearestTargets = new ArrayList<>();

        for (LivingUnit target : targets) {
            if (!isEnemy(wizard.getFaction(), target)) {
                continue;
            }

            if (abs(wizard.getX() - target.getX()) > game.getWizardCastRange() * 2) continue;
            if (abs(wizard.getY() - target.getY()) > game.getWizardCastRange() * 2) continue;

            double distance = wizard.getDistanceTo(target);

            if (distance < wizard.getCastRange()) {
                nearestTargets.add(target);
            }
        }

        return nearestTargets
                .stream()
                .min(Comparator.comparingInt(LivingUnit::getLife));
    }
}

File: src/main/kotlin/FindHelper.kt
import MapHelper.findHelper
import model.*

import java.util.*
import java.util.function.Predicate
import java.util.stream.Collectors

import java.lang.StrictMath.abs

class FindHelper(private val world: World, private val game: Game, private val wizard: Wizard) {

    private val skillHelper: SkillHelper = SkillHelper(game, wizard)

    fun getAllUnits(withTrees: Boolean, onlyEnemy: Boolean, onlyNearest: Boolean, withNeutrals: Boolean): List<LivingUnit> {
        val cacheKey = listOf(withTrees, onlyEnemy, onlyNearest, withNeutrals)
        if (allUnitsCache[cacheKey] != null) return allUnitsCache[cacheKey]!!

        val units = mutableListOf<LivingUnit>()

        units.addAll(getAllWizards(onlyEnemy, onlyNearest))
        units.addAll(getAllBuldings(onlyEnemy))
        units.addAll(getAllMinions(onlyEnemy, onlyNearest, withNeutrals))

        if (withTrees)
            units.addAll(Arrays.asList(*world.getTrees()))

        allUnitsCache.put(cacheKey, units)

        return units
    }

    fun getAllMovingUnits(onlyEnemy: Boolean, onlyNearest: Boolean, withNeutrals: Boolean = false): List<LivingUnit> {
        val cacheKey = listOf(onlyEnemy, onlyNearest, withNeutrals)
        if (allMovingUnitsCache[cacheKey] != null) return allMovingUnitsCache[cacheKey]!!

        val units = ArrayList<LivingUnit>()

        units.addAll(getAllWizards(onlyEnemy, onlyNearest))
        units.addAll(getAllMinions(onlyEnemy, onlyNearest, withNeutrals))

        allMovingUnitsCache.put(cacheKey, units)

        return units
    }

    fun getAllWizards(onlyEnemy: Boolean, onlyNearest: Boolean): List<Wizard> {
        val cacheKey = Arrays.asList(onlyEnemy, onlyNearest)

        if (allWizardsCache[cacheKey] != null) return allWizardsCache[cacheKey]!!

        val newUnins = world.getWizards()
                .filter { wizard -> !wizard.isMe }
                .filter { filterLivingUnits(it, onlyEnemy, onlyNearest) }

        allWizardsCache.put(cacheKey, newUnins)

        return newUnins
    }

    private fun filterLivingUnits(unit: LivingUnit, onlyEnemy: Boolean, onlyNearest: Boolean, withNeutrals: Boolean = false): Boolean {
        return !onlyEnemy || isEnemy(wizard.faction, unit) || (unit.faction == Faction.NEUTRAL && withNeutrals)
                && (!onlyNearest || abs(unit.x - wizard.x) < game.wizardCastRange * 2)
                && (!onlyNearest || abs(unit.y - wizard.y) < game.wizardCastRange * 2)
    }

    fun getAllBuldings(onlyEnemy: Boolean): List<Building> {
        val cacheKey = onlyEnemy

        if (allBuldings[cacheKey] != null) return allBuldings[cacheKey]!!

        val mapSize = game.mapSize

        val enemyNotVisionBuildings = world.getBuildings().filter {
            !isEnemy(wizard.faction, it)
        }.map {
            Building(
                    id = -1,
                    x = mapSize - it.x,
                    y = mapSize - it.y,
                    speedX = 0.0,
                    speedY = 0.0,
                    angle = it.angle,
                    faction = if (wizard.faction == Faction.ACADEMY) Faction.RENEGADES else Faction.ACADEMY,
                    radius = it.radius,
                    life = 100,
                    maxLife = it.life,
                    statuses = emptyArray(),
                    type = it.type,
                    visionRange = it.visionRange,
                    attackRange = it.attackRange,
                    cooldownTicks = 0,
                    remainingActionCooldownTicks = 0,
                    damage = 100500)
        }

        val (enemyBuldings, notEnemyBuildings) = world.getBuildings().toList()
                .partition { isEnemy(wizard.faction, it) }

        val resultEnemyBuldings = enemyNotVisionBuildings
                .filter { building ->
                    enemyBuldings.none { it.getDistanceTo(building) <= MIN_CLOSEST_DISTANCE }
                    MapHelper.deadGuardTowers.values.none { it.getDistanceTo(building) <= MIN_CLOSEST_DISTANCE }
                } + enemyBuldings

        val newUnits = if (onlyEnemy) resultEnemyBuldings else resultEnemyBuldings + notEnemyBuildings

        allBuldings.put(cacheKey, newUnits)

        return newUnits
    }

    fun getAllMinions(onlyEnemy: Boolean, onlyNearest: Boolean, withNeutrals: Boolean = false): List<Minion> {
        val cacheKey = Arrays.asList(onlyEnemy, onlyNearest, withNeutrals)

        if (allMinions[cacheKey] != null) return allMinions[cacheKey]!!

        val newUnits = world.getMinions()
                .filter { filterLivingUnits(it, onlyEnemy, onlyNearest, withNeutrals) }

        allMinions.put(cacheKey, newUnits)

        return newUnits
    }

    fun getAllMovingNeutrals(): List<Minion> {
        if (allMovingNeutrals != null) return allMovingNeutrals!!

        val neutrals = world.getMinions()
                .filter { filterLivingUnits(it, onlyEnemy = false, onlyNearest = true, withNeutrals = true) }
                .filter { it.faction == Faction.NEUTRAL }
                .filter { it.speedX > 0 || it.speedY > 0 }

        allMovingNeutrals = neutrals

        return neutrals
    }

    fun getAllTrees(): List<Tree> {
        if (allTrees != null) return allTrees!!

        val newUnits = world.getTrees()
                .filter { filterLivingUnits(it, false, true) }

        allTrees = newUnits

        return newUnits
    }

    fun isEnemy(self: Faction, unit: LivingUnit): Boolean {
        return self !== unit.faction && unit.faction !== Faction.NEUTRAL
    }

    fun getNearestTarget(): LivingUnit? {
        val nearestWizard = getNearestTarget(findHelper.getAllWizards(onlyEnemy = true, onlyNearest = true))

        if (nearestWizard != null) return nearestWizard

        val nearestBuilding = getNearestTarget(findHelper.getAllBuldings(onlyEnemy = true))

        if (nearestBuilding != null) return nearestBuilding

        val nearestMinion = getNearestTarget(findHelper.getAllMinions(onlyEnemy = true, onlyNearest = true))

        if (nearestMinion != null) return nearestMinion

        val nearestMovingNeutrals = getNearestTarget(getAllMovingNeutrals())

        if (nearestMovingNeutrals != null) return nearestMovingNeutrals

        return null
    }

    fun getNearestTarget(targets: List<LivingUnit>, minionType: MinionType? = null): LivingUnit? {
        val nearestTargets = ArrayList<LivingUnit>()

        for (target in targets) {
            if (minionType != null && target is Minion && target.type != minionType) continue

            if (abs(wizard.x - target.x) > game.wizardCastRange * CAST_RANGE_FACTOR) continue
            if (abs(wizard.y - target.y) > game.wizardCastRange * CAST_RANGE_FACTOR) continue

            val distance = wizard.getDistanceTo(target)

            if (distance <= wizard.castRange ||
                    (target is Wizard && skillHelper.isFirebollActive() && distance - target.radius <= wizard.castRange + game.fireballExplosionMaxDamageRange * 2)) {
                nearestTargets.add(target)
            }
        }

        if (nearestTargets.isNotEmpty() && nearestTargets[0] is Wizard)
            return nearestTargets
                    .minBy { it.life }
        else
            return nearestTargets
                    .minBy { it.getDistanceTo(wizard) }
    }

    companion object {

        private var allUnitsCache: MutableMap<List<Boolean>, List<LivingUnit>> = HashMap()
        private var allMovingUnitsCache: MutableMap<List<Boolean>, List<LivingUnit>> = HashMap()
        private var allWizardsCache: MutableMap<List<Boolean>, List<Wizard>> = HashMap()
        private var allBuldings: MutableMap<Boolean, List<Building>> = HashMap()
        private var allMinions: MutableMap<List<Boolean>, List<Minion>> = HashMap()
        private var allMovingNeutrals: List<Minion>? = null
        private var allTrees: List<Tree>? = null

        fun clearCache() {
            allUnitsCache = HashMap<List<Boolean>, List<LivingUnit>>()
            allMovingUnitsCache = HashMap<List<Boolean>, List<LivingUnit>>()
            allWizardsCache = HashMap<List<Boolean>, List<Wizard>>()
            allBuldings = HashMap<Boolean, List<Building>>()
            allMinions = HashMap<List<Boolean>, List<Minion>>()
            allTrees = null
        }

        val CAST_RANGE_FACTOR: Double = 1.3

        val MIN_CLOSEST_DISTANCE: Double = 15.0
    }
}
